File: UTESA/DCE 5 Estrellas/dce2005/Estrella1/DCE1_Ejemplos/src/VB/ReglasNegocio/CatalogoProductos.vb
Imports DCE05.Ejemplos.EstrellaUno.AccesoDatos
Imports System.Data.Common

''' <summary>
''' Representa el catálogo de productos del sistema.
''' </summary>
Public Class CatalogoProductos

    ''' <summary>
    ''' Obtiene la lista de productos del sistema.
    ''' </summary>
    ''' <returns>La lista de productos.</returns>
    ''' <exception cref="ReglasNegocioException">Si ocurre un error de negocio.</exception>
    Public Function ObtenerProductos() As List(Of Producto)
        Dim Productos As List(Of Producto) = New List(Of Producto)()
        Try
            Dim Sql As String = "SELECT Codigo, Descripcion, Precio FROM Productos"
            Dim Db As BaseDatos = New BaseDatos()
            Db.Conectar()
            Db.CrearComando(Sql)
            Dim Datos As DbDataReader = Db.EjecutarConsulta()

            Dim p As Producto = Nothing
            While Datos.Read()
                Try
                    p = New Producto(Datos.GetInt32(0), Datos.GetString(1), Datos.GetDouble(2))
                    Productos.Add(p)
                Catch ex As InvalidCastException
                    Throw New ReglasNegocioException("Los tipos no coinciden.", ex)
                Catch ex As DataException
                    Throw New ReglasNegocioException("Error de ADO.NET.", ex)
                End Try
            End While
            Datos.Close()
            Db.Desconectar()
        Catch ex As BaseDatosException
            Throw New ReglasNegocioException("Error al acceder a la base de datos para obtener los productos.")
        Catch ex As ReglasNegocioException
            Throw New ReglasNegocioException("Error a obtener los productos.")
        End Try
        Return Productos
    End Function

    ''' <summary>
    ''' Obtiene un producto en base a su código.
    ''' </summary>
    ''' <param name="Codigo">El código del producto.</param>
    ''' <returns>El producto encontrado.</returns>
    ''' <exception cref="ReglasNegocioException">Si ocurre un error de negocio.</exception>
    Public Function ObtenerProducto(ByVal Codigo As Integer) As Producto
        Dim MiProducto As Producto = Nothing
        Try
            If (Codigo <= 0) Then
                Throw New ReglasNegocioException("El código es inválido.")
            End If

            Dim Sql As String = "SELECT Codigo, Descripcion, Precio FROM Productos WHERE Codigo=@codigo"
            Dim Db As BaseDatos = New BaseDatos()
            Db.Conectar()
            Db.CrearComando(Sql)
            Db.AsignarParametroEntero("@codigo", Codigo)
            Dim Datos As DbDataReader = Db.EjecutarConsulta()

            While Datos.Read()
                Try
                    MiProducto = New Producto(Datos.GetInt32(0), Datos.GetString(1), Datos.GetDouble(2))
                Catch Ex As InvalidCastException
                    Throw New ReglasNegocioException("Los tipos no coinciden.", Ex)
                Catch Ex As DataException
                    Throw New ReglasNegocioException("Error de ADO.NET.", Ex)
                End Try
            End While

            Datos.Close()
            Db.Desconectar()
        Catch Ex As BaseDatosException
            Throw New ReglasNegocioException("Error al acceder a la base de datos para obtener los productos.")
        Catch Ex As ReglasNegocioException
            Throw New ReglasNegocioException("Error a obtener los productos. " + Ex.Message)
        End Try
        Return MiProducto
    End Function
End Class

File: UTESA/DCE 5 Estrellas/dce2005/Estrella1/DCE1_Ejemplos/src/CS/ReglasNegocio/CatalogoProductos.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Data.Common;
using System.Data;

using DCE05.Ejemplos.EstrellaUno.AccesoDatos;

namespace DCE05.Ejemplos.EstrellaUno.ReglasNegocio {
    
    /// <summary>
    /// Representa el catálogo de productos del sistema.
    /// </summary>
    public class CatalogoProductos {

        /// <summary>
        /// Obtiene la lista de productos del sistema.
        /// </summary>
        /// <returns>La lista de productos.</returns>
        /// <exception cref="ReglasNegocioException">Si ocurre un error de negocio.</exception>
        public List<Producto> ObtenerProductos() {
            List<Producto> productos = new List<Producto>();
            try {
                string sql = "SELECT Codigo, Descripcion, Precio FROM Productos";
                BaseDatos db = new BaseDatos();
                db.Conectar();
                db.CrearComando(sql);
                DbDataReader datos = db.EjecutarConsulta();

                Producto p = null;
                while (datos.Read()) {
                    try {
                        p = new Producto(datos.GetInt32(0), datos.GetString(1), datos.GetDouble(2));
                        productos.Add(p);
                    } catch (InvalidCastException ex) {
                        throw new ReglasNegocioException("Los tipos no coinciden.", ex);
                    } catch (DataException ex) {
                        throw new ReglasNegocioException("Error de ADO.NET.", ex);
                    }
                }
                datos.Close();
                db.Desconectar();
            } catch (BaseDatosException) {
                throw new ReglasNegocioException("Error al acceder a la base de datos para obtener los productos.");
            } catch (ReglasNegocioException) {
                throw new ReglasNegocioException("Error a obtener los productos.");
            }
            return productos;
        }

        /// <summary>
        /// Obtiene un producto en base a su código.
        /// </summary>
        /// <param name="codigo">El código del producto.</param>
        /// <returns>El producto encontrado.</returns>
        /// <exception cref="ReglasNegocioException">Si ocurre un error de negocio.</exception>
        public Producto ObtenerProducto(int codigo) {
            Producto producto = null;

            try {

                if (codigo <= 0) {
                    throw new ReglasNegocioException("El código es inválido.");
                }

                string sql = "SELECT Codigo, Descripcion, Precio FROM Productos WHERE Codigo=@codigo";
                BaseDatos db = new BaseDatos();
                db.Conectar();
                db.CrearComando(sql);
                db.AsignarParametroEntero("@codigo", codigo);
                DbDataReader datos = db.EjecutarConsulta();

                while (datos.Read()) {
                    try {
                        producto = new Producto(datos.GetInt32(0), datos.GetString(1), datos.GetDouble(2));
                    } catch (InvalidCastException ex) {
                        throw new ReglasNegocioException("Los tipos no coinciden.", ex);
                    } catch (DataException ex) {
                        throw new ReglasNegocioException("Error de ADO.NET.", ex);
                    }
                }
                datos.Close();
                db.Desconectar();
            } catch (BaseDatosException) {
                throw new ReglasNegocioException("Error al acceder a la base de datos para obtener los productos.");
            } catch (ReglasNegocioException ex) {
                throw new ReglasNegocioException("Error a obtener los productos. " + ex.Message);
            }
            return producto;
        }

    }
}
